Source/HTTP/Curl/CurlEasyRequest.cpp

```diff
@@ -0,0 +1,345 @@
+        return S_OK;
+    }
+    else if (maxSize >= bodySize - m_bodyCopied)
+    {
+        // copy everything
+        toCopy = bodySize - m_bodyCopied;
+    }
+    else
+    {
+        // copy as much as we can
+        toCopy = maxSize;
+    }
+
+    void const* startCopyFrom = body + m_bodyCopied;
+    assert(startCopyFrom < body + bodySize);
+
+    m_bodyCopied += toCopy;
+    assert(m_bodyCopied <= bodySize);
+
+    assert(toCopy <= maxSize);
+    memcpy(buffer, startCopyFrom, toCopy);
+
+    bytesCopied = toCopy;
+    return S_OK;
+}
+
+size_t CurlEasyRequest::ReadCallback(char* buffer, size_t size, size_t nitems, void* context) noexcept
+{
+    HC_TRACE_VERBOSE(HTTPCLIENT, "CurlEasyRequest::ReadCallback: reading body data (%llu items of size %llu)", nitems, size);
+
+    auto request = static_cast<CurlEasyRequest*>(context);
+
+    size_t bufferSize = size * nitems;
+    size_t copied = 0;
+    HRESULT hr = request->CopyNextBodySection(buffer, bufferSize, copied);
+    if (FAILED(hr))
+    {
+        return CURL_READFUNC_ABORT;
+    }
+
+    return copied;
+}
+
+size_t CurlEasyRequest::WriteHeaderCallback(char* buffer, size_t size, size_t nitems, void* context) noexcept
+{
+    HC_TRACE_VERBOSE(HTTPCLIENT, "CurlEasyRequest::WriteHeaderCallback: received header (%llu items of size %llu)", nitems, size);
+
+    auto request = static_cast<CurlEasyRequest*>(context);
+
+#if HC_TRACE_INFORMATION_ENABLE
+    if (size * nitems > 2)
+    {
+        HC_TRACE_INFORMATION(HTTPCLIENT, "'%.*s'", size * nitems - 2, buffer); // -2 to avoid printing \r\n
+    }
+#endif
+
+    size_t bufferSize = size * nitems;
+    char const* current = buffer;
+    char const* end = buffer + bufferSize;
+
+    // scan for the end of the header name
+    char const* name = current;
+    size_t nameSize = 0;
+    for (; current < end; ++current)
+    {
+        if (*current == ':')
+        {
+            nameSize = current - buffer;
+            ++current;
+            break;
+        }
+    }
+    if (current == end)
+    {
+        // not a real header, drop it
+        return bufferSize;
+    }
+
+    // skip whitespace
+    for (; current < end && *current == ' '; ++current) // assume that Curl canonicalizes headers
+    {}
+
+    // scan for the end of the header value
+    char const* value = current;
+    size_t valueSize = 0;
+    char const* valueStart = current;
+    for (; current < end; ++current)
+    {
+        if (*current == '\r')
+        {
+            valueSize = current - valueStart;
+            break;
+        }
+    }
+    if (current == end)
+    {
+        // curl should always gives us the new lines at the end of the header
+        assert(false);
+    }
+
+    HRESULT hr = HCHttpCallResponseSetHeaderWithLength(request->m_hcCallHandle, name, nameSize, value, valueSize);
+    assert(SUCCEEDED(hr));
+
+    return bufferSize;
+}
+
+size_t CurlEasyRequest::WriteDataCallback(char* buffer, size_t /*size*/, size_t nmemb, void* context) noexcept
```

Comment:
This should also be size * count as in the other headers.  We only ever return "1" right now as documented by libcurl, but it's still best practice.

Reply:
from https://curl.se/libcurl/c/CURLOPT_WRITEFUNCTION.html
>  ptr points to the delivered data, and the size of that data is nmemb; size is always 1.
To me that says that we don't know what to do with `size` if it is not 1. Maybe we multiply `nmemb` or maybe `nmemb` is premultiplied, who knows? The spec doesn't say and anyone setting `size` to not 1 is breaking the spec, trying to guess how they are breaking it could result in just as many bugs as assuming they aren't breaking it.

Reply:
In theory we want to work with other curl implementations as well.  If the curl spec dictates its always 1 I think we should assume that is true.

Reply:
It's also true in all implementations for all the other callbacks, too.  I think it's not gauranteed for non-HTTP requests which don't matter here, but it's weird to single this one out.

Reply:
looking at the docs, the header callback is also documented to always have 1 for size but not read

https://curl.se/libcurl/c/CURLOPT_HEADERFUNCTION.html
https://curl.se/libcurl/c/CURLOPT_READFUNCTION.html